File: packages/client/src/auto-spawn.ts
/**
 * Auto-spawn helper — 첫 호출 시 데몬이 없으면 백그라운드로 띄운다.
 *
 * 흐름:
 *  1. socketPath에 연결 시도 (짧은 timeout)
 *  2. 실패 → muxd 바이너리 찾아서 `muxd serve` detached spawn
 *  3. 데몬 ready 폴링 (status ping)
 *  4. 호출자에게 connect 가능 시점 알림
 *
 * 바이너리 위치 우선순위:
 *   1. opts.muxdPath
 *   2. process.env.MUXD_BIN
 *   3. PATH에서 `muxd` (where/which)
 *
 * Note: 호출자 프로세스가 죽어도 데몬은 살아남도록 detach.
 */

import { spawn } from "node:child_process";
import net from "node:net";
import { execSync } from "node:child_process";
import { MuxClientError } from "./errors.js";

export interface AutoSpawnOpts {
  socketPath: string;
  /** muxd 바이너리 경로 명시 */
  muxdPath?: string;
  /** spawn 후 데몬 ready 폴링 timeout (ms). 기본 5000 */
  spawnTimeoutMs?: number;
}

/** 데몬이 listening 중인지 확인 — 짧은 connect 시도 */
export async function isDaemonRunning(
  socketPath: string,
  timeoutMs = 300,
): Promise<boolean> {
  return new Promise((resolve) => {
    const sock = net.createConnection(socketPath);
    const timer = setTimeout(() => {
      sock.destroy();
      resolve(false);
    }, timeoutMs);
    sock.on("connect", () => {
      clearTimeout(timer);
      sock.end();
      resolve(true);
    });
    sock.on("error", () => {
      clearTimeout(timer);
      resolve(false);
    });
  });
}

function findMuxdBinary(explicit?: string): string {
  if (explicit) return explicit;
  if (process.env.MUXD_BIN) return process.env.MUXD_BIN;
  // PATH lookup
  try {
    const finder = process.platform === "win32" ? "where" : "which";
    const out = execSync(`${finder} muxd`, { stdio: ["ignore", "pipe", "ignore"] })
      .toString()
      .trim();
    const first = out.split(/\r?\n/)[0]?.trim();
    if (first) return first;
  } catch {}
  throw new MuxClientError(
    "DAEMON_NOT_FOUND",
    "muxd binary not found. Set MUXD_BIN or pass opts.muxdPath, or install @claude-mux/muxd globally.",
  );
}

/** muxd serve를 detached로 띄우고 ready 될 때까지 대기. */
export async function spawnDaemon(opts: AutoSpawnOpts): Promise<void> {
  const bin = findMuxdBinary(opts.muxdPath);
  const timeoutMs = opts.spawnTimeoutMs ?? 5000;
  // Windows에서 muxd는 .cmd shim일 수 있음 — shell: true로 안전하게
  const child = spawn(bin, ["serve"], {
    detached: true,
    stdio: "ignore",
    shell: process.platform === "win32",
  });
  child.unref(); // 호출자가 죽어도 데몬 생존

  const startedAt = Date.now();
  while (Date.now() - startedAt < timeoutMs) {
    if (await isDaemonRunning(opts.socketPath, 200)) return;
    await new Promise((r) => setTimeout(r, 200));
  }
  throw new MuxClientError(
    "CONNECT_FAILED",
    `muxd spawned but not responding on ${opts.socketPath} within ${timeoutMs}ms`,
  );
}

File: packages/client-py/src/claude_mux/auto_spawn.py
"""
Auto-spawn — 데몬이 없으면 `muxd serve`를 detached로 띄움.

위치 우선순위:
  1. opts.muxd_path
  2. MUXD_BIN env
  3. PATH lookup (`where`/`which`)
"""

from __future__ import annotations

import os
import shutil
import socket
import subprocess
import sys
import time
from typing import Optional

from .errors import MuxClientError


def is_daemon_running(socket_path: str, timeout_ms: int = 300) -> bool:
    """짧은 connect 시도로 데몬 listening 여부 확인."""
    deadline = time.monotonic() + timeout_ms / 1000.0
    if sys.platform == "win32":
        # Windows Named pipe — 존재 확인은 CreateFile 시도가 가장 확실
        try:
            import pywintypes  # type: ignore
            import win32file  # type: ignore

            while time.monotonic() < deadline:
                try:
                    handle = win32file.CreateFile(
                        socket_path,
                        win32file.GENERIC_READ,
                        0,
                        None,
                        win32file.OPEN_EXISTING,
                        0,
                        None,
                    )
                    handle.Close()
                    return True
                except pywintypes.error:
                    time.sleep(0.05)
            return False
        except ImportError:
            return False
    # POSIX
    while time.monotonic() < deadline:
        try:
            sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
            sock.settimeout(0.1)
            sock.connect(socket_path)
            sock.close()
            return True
        except OSError:
            time.sleep(0.05)
    return False


def _find_muxd_binary(explicit: Optional[str] = None) -> str:
    if explicit:
        return explicit
    env = os.environ.get("MUXD_BIN")
    if env:
        return env
    found = shutil.which("muxd")
    if found:
        return found
    raise MuxClientError(
        "DAEMON_NOT_FOUND",
        "muxd binary not found. Set MUXD_BIN, pass muxd_path, "
        "or install @claude-mux/muxd globally.",
    )


def spawn_daemon(
    socket_path: str,
    muxd_path: Optional[str] = None,
    spawn_timeout_ms: int = 5000,
) -> None:
    """muxd serve를 detached로 spawn 후 ready 폴링."""
    binary = _find_muxd_binary(muxd_path)
    # Windows에서 .cmd shim일 수 있으므로 shell=True
    if sys.platform == "win32":
        subprocess.Popen(
            [binary, "serve"],
            shell=True,
            creationflags=subprocess.DETACHED_PROCESS | subprocess.CREATE_NEW_PROCESS_GROUP,
            stdin=subprocess.DEVNULL,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            close_fds=True,
        )
    else:
        subprocess.Popen(
            [binary, "serve"],
            start_new_session=True,
            stdin=subprocess.DEVNULL,
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            close_fds=True,
        )
    deadline = time.monotonic() + spawn_timeout_ms / 1000.0
    while time.monotonic() < deadline:
        if is_daemon_running(socket_path, timeout_ms=200):
            return
        time.sleep(0.2)
    raise MuxClientError(
        "CONNECT_FAILED",
        f"muxd spawned but not responding on {socket_path} within {spawn_timeout_ms}ms",
    )
